libraries/Microsoft.Bot.Connector/Authentication/MsalAppCredentials.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Identity.Client;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    /// <summary>
+    /// An authentication class that implements <see cref="IAuthenticator"/>, used to acquire tokens for outgoing messages to the channels.
+    /// </summary>
+    public class MsalAppCredentials : AppCredentials, IAuthenticator
+    {
+        /// <summary>
+        /// An empty set of credentials.
+        /// </summary>
+        public static readonly MsalAppCredentials Empty = new MsalAppCredentials(clientApplication: null, appId: null);
+
+        // Semaphore to control concurrency while refreshing tokens from MSAL.
+        // Whenever a token expires, we want only one request to retrieve a token.
+        // Cached requests take less than 0.1 millisecond to resolve, so the semaphore doesn't hurt performance under load tests
+        // unless we have more than 10,000 requests per second, but in that case other things would break first.
+        private static SemaphoreSlim tokenRefreshSemaphore = new SemaphoreSlim(1, 1);
```

Comment:
The logic around refresh and exception handling here is the same as we used with ADAL, but with some simplifications because error handling is much simpler in MSAL. So not creating anything new here, but reusing the pattern that has worked for years.